### PR TITLE
[fix] Self-closing tags are not supported as top-level rows

### PR DESCRIPTION
**Pull request by hamzaezzi**
fix bug of Self-closing in top-level rows

test with :

xml file content 

```
<?xml version='1.0' encoding='UTF-8'?>
<parentNode version="0.6">
  <child id="122944" version="11"/>
  <child id="122945" version="11"/>
  <child id="122946" version="12"/>
  <child id="122954" version="11">
    <tag k="toto" v="titi"/>
    <tag k="toto2" v="titi2"/>
  </child>  
</parentNode>
```

Java code

```
DataFrame lines = sqlContext
            .read()
            .format("com.databricks.spark.xml")
            .option("rowTag", "child")
            .load(filePathNodes)
lines.printSchema();
lines.show();
```

### CODE REVIEW

**Comment by HyukjinKwon** [comment]
@hamzaezzi It seems tests are being failed. Could you fix this to pass them?

BTW, I am a bit worried of the performance due to the use of regular expression. Since `XmlInputFormat.next()` is a critical path, this generally degrade the performance. Would there be another way to avoid the use of regular expression?

Also, I like some points you made (e.g. some styles and the approach you took) although we should fix some according to [databricks/scala-style-guide](https://github.com/databricks/scala-style-guide).

**Comment by hamzaezzi** [comment]
can you give me your use case becaue it works fine for me

for "regular expression" you can replace it with a convertion of "buffer.getData" to xml object and parse it to check if it containt others tags but i'm not sure that it will be more efficient...

**Comment by HyukjinKwon** [comment]
Could you please try this command below?

```bash
./sbt/sbt test scalastyle "test:scalastyle"
```

This will run tests existing in https://github.com/databricks/spark-xml/blob/master/src/test/scala/com/databricks/spark/xml/XmlSuite.scala and will also check basic Scala style as configured in https://github.com/databricks/spark-xml/blob/master/scalastyle-config.xml

and.. actually that is the reason why I could not submitted a PR for this by now..

**Comment by HyukjinKwon** [comment]
ping @hamzaezzi Do you think we should close this for now and open another if we have a better idea on this?

**Comment by hamzaezzi** [comment]
Sorry for the delay
I had developed a new version that passed the most tests. I will push my code next week, actually I'm on holiday and I don't have my computer :)  
Can you wait next week to validate the new version?

**Comment by HyukjinKwon** [comment]
Glad to know. Sure, I am not pushing you but just wondering what we should do with this :).

**Comment by pchen0629** [comment]
any updates?

**Comment by HyukjinKwon** [comment]
@hamzaezzi It'd be great if you rebase this PR for the conflicts. FYI, you could refer this comment for updating your PR - https://github.com/databricks/spark-xml/pull/174#issuecomment-244638465

**Comment by hamzaezzi** [comment]
hello,

i had merged my fork with the 0.4.0

i don't know what do you have modified exactly, i have the exact lines but with null value !!

i have tested it with spark 1.6.1, should i use spark 2 ?

    <?xml version='1.0' encoding='UTF-8'?>
    <osm version="0.6" generator="Osmosis 0.45">
      <node id="482600400" version="4" timestamp="2012-12-01T00:14:36Z" uid="987535" user="blieusong" changeset="14106152" lat="48.8382618" lon="2.6562604">
        <tag k="source" v="cadastre-dgi-fr source : Direction Générale des Impôts - Cadastre. Mise à jour : 2009"/>
      </node>
      <node id="482541161" version="5" timestamp="2012-10-10T23:07:45Z" uid="870861" user="botdidier2020" changeset="13448298" lat="48.853342" lon="2.3408242"/>
      <node id="482541166" version="4" timestamp="2011-10-04T01:38:22Z" uid="172867" user="AmiFritz" changeset="9465758" lat="48.8530827" lon="2.340525"/>
      <node id="482541169" version="4" timestamp="2011-10-04T01:38:22Z" uid="172867" user="AmiFritz" changeset="9465758" lat="48.8532261" lon="2.3406202"/>
      <node id="482541172" version="4" timestamp="2011-10-04T01:38:22Z" uid="172867" user="AmiFritz" changeset="9465758" lat="48.8532463" lon="2.3406016"/>
      <node id="482541175" version="4" timestamp="2011-10-04T01:38:22Z" uid="172867" user="AmiFritz" changeset="9465758" lat="48.8532816" lon="2.3401526"/>
      <node id="482541178" version="4" timestamp="2011-10-04T01:38:22Z" uid="172867" user="AmiFritz" changeset="9465758" lat="48.8532209" lon="2.3401046"/>
      <node id="482600393" version="4" timestamp="2012-12-01T00:14:36Z" uid="987535" user="blieusong" changeset="14106152" lat="48.8382099" lon="2.6563661">
        <tag k="source" v="cadastre-dgi-fr source : Direction Générale des Impôts - Cadastre. Mise à jour : 2009"/>
      </node>
      <node id="482600400" version="4" timestamp="2012-12-01T00:14:36Z" uid="987535" user="blieusong" changeset="14106152" lat="48.8382618" lon="2.6562604">
        <tag k="source" v="cadastre-dgi-fr source : Direction Générale des Impôts - Cadastre. Mise à jour : 2009"/>
      </node>
    </osm>

    +----+----+----+------+
    | @id|@lat|@lon|   tag|
    +----+----+----+------+
    |null|null|null|[null]|
    |null|null|null|  null|
    |null|null|null|  null|
    |null|null|null|  null|
    |null|null|null|  null|
    |null|null|null|  null|
    |null|null|null|  null|
    |null|null|null|[null]|
    |null|null|null|[null]|
    +----+----+----+------+

can you check my fork

https://github.com/hamzaezzi/spark-xml



cordially

**Comment by hamzaezzi** [comment]
can you give me the method when we parse xml tags and transform it to row, this is was changed between 0.3.x and 0.4.0 version

**Comment by hamzaezzi** [comment]
yes, it's working now,

i got the null value because i specified the schema like old version 0.3.3 with prefix "@" for attributes :( 

now i tested it with prefix "_" and it works like a charm

can you confirm that? to valide my merge 

cordially

**Comment by tanny90** [comment]
can you guys please fix this bug also in 3.x branch with spark less than 2.0?

**Comment by HyukjinKwon** [comment]
@tanny90 I will backport importance changes into branch.3.x. I hope the travis tests were passed here.

**Comment by HyukjinKwon** [comment]
@hamzaezzi Could you please take a look for https://travis-ci.org/databricks/spark-xml/builds/162042202? We should test them against 1.3.x+. You could run the tests such as

```
sbt -Dspark.testVersion=1.3.0 test
sbt -Dspark.testVersion=1.6.0 test
```

and 

```
sbt scalastyle
```

for style checking.

**Comment by HyukjinKwon** [comment]
I will take a look at my best as soon as the travis becomes green.

**Comment by leniel** [comment]
Guys,

I'm trying to load the StackOverflow data dump (Users.xml)... it has this format with self-closing row tags:

```
<?xml version="1.0" encoding="utf-8"?>
<users>
  <row Id="-1" Reputation="1" CreationDate="2008-07-31T00:00:00.000" DisplayName="Community" LastAccessDate="2008-08-26T00:16:53.810" WebsiteUrl="http://meta.stackexchange.com/" Location="on the server farm" AboutMe="&lt;p&gt;Hi, I'm not really a person.&lt;/p&gt;&#xA;&#xA;&lt;p&gt;I'm a background process that helps keep this site clean!&lt;/p&gt;&#xA;&#xA;&lt;p&gt;I do things like&lt;/p&gt;&#xA;&#xA;&lt;ul&gt;&#xA;&lt;li&gt;Randomly poke old unanswered questions every hour so they get some attention&lt;/li&gt;&#xA;&lt;li&gt;Own community questions and answers so nobody gets unnecessary reputation from them&lt;/li&gt;&#xA;&lt;li&gt;Own downvotes on spam/evil posts that get permanently deleted&lt;/li&gt;&#xA;&lt;li&gt;Own suggested edits from anonymous users&lt;/li&gt;&#xA;&lt;li&gt;&lt;a href=&quot;http://meta.stackexchange.com/a/92006&quot;&gt;Remove abandoned questions&lt;/a&gt;&lt;/li&gt;&#xA;&lt;/ul&gt;&#xA;" Views="649" UpVotes="170442" DownVotes="725896" AccountId="-1" />
```

Using:

```
df = sqlContext.read
      .format("xml")
      .option("rowTag", "row")
      .load("data/Users.xml") // Reading from the local disk

df.printSchema()
```

As the output I get:

```root```

It's failing... :(

The version I'm using is this one:

https://mvnrepository.com/artifact/com.databricks/spark-xml_2.10/0.4.1

Can you please tell me if there's a fix for this already?


Thank you very much!

**Comment by HyukjinKwon** [comment]
@leniel Could you try to do it with `users` as `rowTag` and use `explode(..)` as described in https://github.com/databricks/spark-xml/issues/141#issuecomment-234835577 please? Let us talk in https://github.com/databricks/spark-xml/issues/92 if you face more problems with this.

**Comment by revolutionisme** [comment]
Hi @HyukjinKwon,
Any idea when this is going to get included in the spark-xml stack? I mean I can read XML contents using explode as described in #141 but that wouldn't work when i am reading a huge XML (for eg. stackoverflow data dump) file as it would try to read everything in a dataframe and it crashes because "Requested array size exceeds VM limit"

**Comment by Treydone** [comment]
Hi @HyukjinKwon, is there any news about this issue? I have the same problem described by @revolutionisme, I'm working on a huge XML and the solution using the explode crashes :(

**Comment by denniseffing** [comment]
Hi @HyukjinKwon,
I just found out about your Apache Spark library and immediately ran into this issue. It would be much appreciated if you'd work on this PR and merge it in the near future.

**Comment by HyukjinKwon** [comment]
Yea, it is tricky. I will probably try to have some time to give a shot for this one.

**Comment by DaniSancas** [comment]
Hi,

I ran into this problem too. It would be interesting to solve this issue, I think many of us will appreciate it!

**Comment by sdikby** [comment]
Ping @HyukjinKwon @hamzaezzi 
It would be very much appreciated if this problem is solved.

**Comment by hamzaezzi** [comment]
I will do it this week :) i hope there will not be a new version in the meantime :/

**Comment by DaniSancas** [comment]
Hope you can do it, we'll love you so much!

File: src/main/scala/com/databricks/spark/xml/XmlInputFormat.scala [hunk 1]
```diff
@@ -75,10 +78,13 @@ private[xml] class XmlRecordReader extends RecordReader[LongWritable, Text] {
       Charset.forName(conf.get(XmlInputFormat.ENCODING_KEY, XmlOptions.DEFAULT_CHARSET))
     startTag = conf.get(XmlInputFormat.START_TAG_KEY).getBytes(charset)
     endTag = conf.get(XmlInputFormat.END_TAG_KEY).getBytes(charset)
+    endEmptyTag = "/>".getBytes(charset)
     space = " ".getBytes(charset)
     angleBracket = ">".getBytes(charset)
+    tagPattern = ("\\"+new String(startTag).substring(0,new String(startTag).length-1)+"([^\\>]*)").r
```
Comment (HyukjinKwon) [inline src/main/scala/com/databricks/spark/xml/XmlInputFormat.scala]: This pattern matching will degrade performance a lot ..

File: src/main/scala/com/databricks/spark/xml/XmlInputFormat.scala [hunk 2]
```diff
@@ -137,13 +143,23 @@ private[xml] class XmlRecordReader extends RecordReader[LongWritable, Text] {
    * @return whether it reads successfully
    */
   private def next(key: LongWritable, value: Text): Boolean = {
-    if (readUntilStartElement()) {
+    if (readUntilMatch(startTag, withinBlock = false, force = false)) {
```
Comment (HyukjinKwon) [inline src/main/scala/com/databricks/spark/xml/XmlInputFormat.scala]: I guess here is the very core in this library, performance sensitive and affect all the functionailities .. it'd be nicer if we have tests.

File: src/main/scala/com/databricks/spark/xml/XmlInputFormat.scala [hunk 3]
```diff
@@ -137,13 +143,23 @@ private[xml] class XmlRecordReader extends RecordReader[LongWritable, Text] {
...
       try {
         buffer.write(currentStartTag)
-        if (readUntilEndElement()) {
-          key.set(filePosition.getPos)
-          value.set(buffer.getData, 0, buffer.getLength)
-          true
+        if (readUntilMatch(endTag, withinBlock = true, force = false)) {
```
Comment (HyukjinKwon) [inline src/main/scala/com/databricks/spark/xml/XmlInputFormat.scala]: There is a style guide here - https://github.com/databricks/scala-style-guide

**Comment by HyukjinKwon** [comment]
@hamzaezzi, I am sorry for this delay. Would you maybe have some time to proceed this further please?

**Comment by kvalev** [comment]
Is there any way to help out? It is a bit annoying to work around this issue :/

**Comment by BioQwer** [comment]
Hello, when this PR will be merged?

**Comment by srowen** [comment]
This should be fixed by https://github.com/databricks/spark-xml/pull/352

**Pull request closed without merging**